In [15]:
#import library
import pandas as pd

#refer to resources file
pymoli_data = pd.read_csv("./Resources/purchase_data.csv")


In [16]:
#create a DataFrame that has the unique list of names of the players
unique_players_df = pymoli_data.drop_duplicates(subset="SN")

#find total number of players, by finding the length of the unique players DataFrame
total_players= [{"Total Players":len(unique_players_df)}]

#place into a DataFrame
total_players_df=pd.DataFrame(total_players)

#print DataFrame
total_players_df

Total Players
0            576

In [17]:
#PURCHASING ANALYSIS

#number of unique items
unique_items = pymoli_data["Item ID"].nunique()

#average price
avg_price = pymoli_data["Price"].mean()
avg_price

#number of purchases
total_purchases = pymoli_data["Purchase ID"].count()
total_purchases

#total revenue
total_rev = pymoli_data["Price"].sum()
total_rev

#place all values into dictionary
pymoli_purch_analysis = [{"Number of Unique Items":unique_items,
                          "Average Price":avg_price,
                          "Number of Purchases":total_purchases,
                          "Total Revenue":total_rev}]

#move dictionary into a DataFrame
pymoli_purch_analysis_df=pd.DataFrame(pymoli_purch_analysis)

#format the pricing values to show as currency values
pymoli_purch_analysis_df["Average Price"] = pymoli_purch_analysis_df["Average Price"].map('${:,.2f}'.format)
pymoli_purch_analysis_df["Total Revenue"] = pymoli_purch_analysis_df["Total Revenue"].map('${:,.2f}'.format)

#print DataFrame
pymoli_purch_analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

In [18]:
#GENDER DEMOGRAPHICS

#count each gender within the unique players DataFrame, then find the percentage that that gender makes up
gender_count = unique_players_df.Gender.value_counts()
gender_percentage = unique_players_df.Gender.value_counts(normalize=True)

#place into a DataFrame
gender_df = pd.DataFrame({"Total Count":gender_count,
              "Percentage of Players":gender_percentage})

#format the percentage as a percentage rather than a number
gender_df["Percentage of Players"] = gender_df["Percentage of Players"].map('{:,.2%}'.format)

#print DataFrame
gender_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

In [19]:
#PURCHASING ANALYSIS (GENDER)

#group data by Gender column
grouped_gender = pymoli_data.groupby('Gender')

#purchase count
gen_purch_count = grouped_gender["Item ID"].count()

#average purchase price
avg_gender_price = grouped_gender['Price'].mean()

#total purchase value
total_gender_price = grouped_gender['Price'].sum()

#average total purchase per person
avg_total_gender = total_gender_price/gender_count

#create DataFrame containing analysis info
gender_purch_analysis_df = pd.DataFrame({"Purchase Count":gen_purch_count,
                          "Average Purchase Price":avg_gender_price,
                          "Total Purchase Value":total_gender_price,
                          "Average Total Purchase per Person":avg_total_gender})

#format purchase values to currency
gender_purch_analysis_df["Average Purchase Price"] = gender_purch_analysis_df["Average Purchase Price"].map('${:,.2f}'.format)
gender_purch_analysis_df["Total Purchase Value"] = gender_purch_analysis_df["Total Purchase Value"].map('${:,.2f}'.format)
gender_purch_analysis_df["Average Total Purchase per Person"] = gender_purch_analysis_df["Average Total Purchase per Person"].map('${:,.2f}'.format)

#print DataFrame
gender_purch_analysis_df

Purchase Count Average Purchase Price  \
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Average Total Purchase per Person  
Female                             $361.94                             $4.47  
Male                             $1,967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

In [20]:
#AGE DEMOGRAPHICS

#create bins and labels
bins = [0,9,14,19,24,29,34,39,100]
labels = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]

#Assign bins to each row
unique_players_df["Age Bracket"] = pd.cut(unique_players_df['Age'],bins,labels=labels,include_lowest=True)  # this creates a SettingWithCopyWarning and I don't know why, it still parses fine but if someone could explain that would be awesome

age_count = unique_players_df["Age Bracket"].value_counts()
age_percent = unique_players_df["Age Bracket"].value_counts(normalize=True)

ages_df = pd.DataFrame({"Total Count":age_count,
              "Percentage of Players":age_percent})

ages_df["Percentage of Players"] = ages_df["Percentage of Players"].map('{:,.2%}'.format)

ages_df.sort_index()

<ipython-input-20-9f9b1a9d5a1c>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  unique_players_df["Age Bracket"] = pd.cut(unique_players_df['Age'],bins,labels=labels,include_lowest=True)  # this creates a SettingWithCopyWarning and I don't know why, it still parses fine but if someone could explain that would be awesome


Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

In [21]:
#PURCHASING ANALYSIS (AGE)

pymoli_data["Age Bracket"] = pd.cut(pymoli_data['Age'],bins,labels=labels,include_lowest=True)

#group data by newly created Age Bracket column
grouped_age = pymoli_data.groupby('Age Bracket')

#purchase count
age_purch_count = grouped_age["Item ID"].count()

#average purchase price
avg_age_price = grouped_age['Price'].mean()

#total purchase value
total_age_price = grouped_age['Price'].sum()

#average total purchase per person
avg_total_age = total_age_price/age_count

#create DataFrame containing analysis info
age_purch_analysis_df = pd.DataFrame({"Purchase Count":age_purch_count,
                          "Average Purchase Price":avg_age_price,
                          "Total Purchase Value":total_age_price,
                          "Average Total Purchase per Person":avg_total_age})

#format purchase values to currency
age_purch_analysis_df["Average Purchase Price"] = age_purch_analysis_df["Average Purchase Price"].map('${:,.2f}'.format)
age_purch_analysis_df["Total Purchase Value"] = age_purch_analysis_df["Total Purchase Value"].map('${:,.2f}'.format)
age_purch_analysis_df["Average Total Purchase per Person"] = age_purch_analysis_df["Average Total Purchase per Person"].map('${:,.2f}'.format)

#print DataFrame
age_purch_analysis_df

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Average Total Purchase per Person  
<10                               $4.54  
10-14                             $3.76  
15-19                             $3.86  
20-24                             $4.32  
25-29                             $3.81  
30-34                             $4.12  
35-39                             $4.76  
40+                               $3.19

In [22]:
#TOP SPENDERS

#group data by SN columns
grouped_SN = pymoli_data.groupby('SN')

#purchase count
SN_purch_count = grouped_SN["Item ID"].count()

#average purchase price
avg_SN_price = grouped_SN['Price'].mean()

#total purchase value
total_SN_price = grouped_SN['Price'].sum()

#create DataFrame
top_spenders_df = pd.DataFrame({"Purchase Count":SN_purch_count,
                          "Average Purchase Price":avg_SN_price,
                          "Total Purchase Value":total_SN_price})

sorted_top_spenders_df = top_spenders_df.sort_values("Total Purchase Value",ascending=False)

#format purchase values to currency
sorted_top_spenders_df["Average Purchase Price"] = sorted_top_spenders_df["Average Purchase Price"].map('${:,.2f}'.format)
sorted_top_spenders_df["Total Purchase Value"] = sorted_top_spenders_df["Total Purchase Value"].map('${:,.2f}'.format)



sorted_top_spenders_df.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

In [23]:
#MOST POPULAR ITEM

#retrieve Item ID, Item Name and Item Price columns
pymoli_items = pymoli_data[["Item ID","Item Name","Price"]]

#group by Item ID and Item Name
grouped_items = pymoli_items.groupby(["Item ID","Item Name"])

#purchase count
items_count = grouped_items.count()
items_count = items_count.loc[:,"Price"]

#item price
item_price = grouped_items.mean()
item_price = item_price.loc[:,"Price"]

#total purchase value
item_purchase_total = grouped_items.sum()
item_purchase_total = item_purchase_total.loc[:,"Price"]

#create DataFrame
items_df = pd.DataFrame({"Purchase Count":items_count,
                          "Item Price":item_price,
                          "Total Purchase Value":item_purchase_total})

#sort  DataFrame by purchase count in descending order
popular_items_df=items_df.sort_values("Purchase Count", ascending=False)

#format purchase values to currency
popular_items_df["Item Price"] = popular_items_df["Item Price"].map('${:,.2f}'.format)
popular_items_df["Total Purchase Value"] = popular_items_df["Total Purchase Value"].map('${:,.2f}'.format)

#print preview of DataFrame
popular_items_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

In [24]:
#MOST PROFITABLE ITEM
#sort previous DataFrame by total purchase value in descending order
profitable_items_df = items_df.sort_values("Total Purchase Value", ascending=False)

#format purchase values to currency
profitable_items_df["Item Price"] = profitable_items_df["Item Price"].map('${:,.2f}'.format)
profitable_items_df["Total Purchase Value"] = profitable_items_df["Total Purchase Value"].map('${:,.2f}'.format)

#print preview of DataFrame
profitable_items_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# What did we find out?

**1.** Heroes of Pymori has had more male players (**84.03%**) make purchases than female (**14.06%**) and other/non-disclosed gendered players (**1.91%**), however female players make higher value purchases per person (**$4.47**)

**2.** The age range between 20-24 have made the most purchases (**365 total**) and generated the highest revenue (**1,114.06**), however the 35-39 bracket make higher value purchases per person (**$4.76**)

**3.** The **"Final Critic"** is the most popular and most profitable item in Heroes of Pymori